### DISCUSSION TITLE
Possible to not allow NaN on NumberBox?

### DISCUSSION THREAD

**Discussion by dbriard** [microsoft/microsoft-ui-xaml#8399] [Q&A] [answered]
Hi all,
Is there a way to not allow NaN value on NumberBox?
I would like to attach a NumberBox to a slider, but I don't want to allow empty value in the NumberBox.
If user enter nothing in the Box, the text and value must be the last valid value.
Is there a way to do that or does it require a custom NumberBox?
Thanks a lot!

**Comment by Trix07** [comment]
The NumberBox provides a NumberFormatter that can be used to alter the value within the NumberBox. Although there are built-in formatters available, creating a custom class that implements both the INumberFormatter2 and INumberParser interfaces may be necessary to achieve the specific functionality you described. This would enable you to check for NaN input values and replace them with a different default value, such as 0.

**Reply by dbriard** [reply]
Hi @Trix07,  
the problem is that if input text is null or empty, the number formatter is never called and Value is set to NaN.
I think this control is designed to allow NaN, but I don't understand that NaN support is not optional. They must be a way I didn't thinked about. For example if you want to bind the control on a property that must not be such as FontSize, or Opacity... 

```
void NumberBox::ValidateInput()
{
    // Validate the content of the inner textbox
    if (auto&& textBox = m_textBox.get())
    {
        const auto text = trim(textBox.Text().data());

        // Handles empty TextBox case, set text to current value
        if (text.empty())
        {
            Value(std::numeric_limits<double>::quiet_NaN());
        }
        else
        {
            // Setting NumberFormatter to something that isn't an INumberParser will throw an exception, so this should be safe
            const auto numberParser = NumberFormatter().as<winrt::INumberParser>();
```

**Reply by Trix07** [reply, marked as answer]
You're right, I wasn't aware of that! Sorry for being misleading.

There is an alternative solution that involves creating a custom `IValueConverter` that operates on the same principle. This converter returns 0 if the value passed to it is either NaN or null. You can then use this converter when binding the NumberBox value to the Slider value.

By using this approach, the NumberBox can still be empty while the Slider will default to 0. However, if it is necessary for the NumberBox to never be empty, then perhaps try to do the validation on the `ValueChanged` event.

**Reply by dbriard** [reply]
> You're right, I wasn't aware of that! Sorry for being misleading.
> 
> There is an alternative solution that involves creating a custom `IValueConverter` that operates on the same principle. This converter returns 0 if the value passed to it is either NaN or null. You can then use this converter when binding the NumberBox value to the Slider value.
> 
> By using this approach, the NumberBox can still be empty while the Slider will default to 0. However, if it is necessary for the NumberBox to never be empty, then perhaps try to do the validation on the `ValueChanged` event.

Thank you @Trix07 I'll give a try to ValueConverter in my UWP apps, as I often need to bind numberboxes to sliders.

In my WPF apps, I reimplemented the NumberBox based on the WinUI code and resolved the issue using a third ValidationMode value:

`NaNAndInvalidInputOverwritten = 2 // (I don't like the name, it is ugly)`

**Suggestion for the Team:** May be in a future version of the control, we could have something like that... or may be allow NaN ONLY if the PlaceholderText is not null...

```
private void ValidateInput()
        {
            if (_textBox == null)
                return;

            var text = _textBox.Text!.Trim();

            if (string.IsNullOrEmpty(text) && ValidationMode != NumberBoxValidationMode.NaNAndInvalidInputOverwritten)
            {
                Value = double.NaN;
            }
```

**Comment by SoggyBottomBoy** [comment]
@Trix07 I would have thought it would be fairly common to want to default NaN to 0? After after couple of months of developing a tool which makes heavy use of the number box I too just realised that if a number box loses focus without any text the values in the view model will result in NaN causing exceptions (as it was not a case I expected I needed to handle).

I don't mind the 3rd validation mode potentially suggest it is `InvalidOrNotFiniteInputOverwritten` ie. will only accept valid and finite values.

**Reply by Trix07** [reply]
I agree, it seems like a common scenario, especially when binding other controls like sliders to a number box. Issues like this are why I think developing applications in WinUI can be both annoying and frustrating. A new property on the number box that defines whether it defaults to 0 instead of NaN would be an easy solution, rather than having to create converters or custom controls to override the behavior.

Hopefully, this and other minor issues/bugs are something that the WinUI team will look into at some point.
